include/rocksdb/utilities/object_registry.h

```diff
@@ -76,29 +131,75 @@ class ObjectLibrary {
     AddEntry(T::Type(), entry);
     return factory;
   }
+
+  // Invokes the registrar function with the supplied arg for this library.
+  void Register(const RegistrarFunc& registrar, const std::string& arg) {
+    registrar(*this, arg);
+  }
+
   // Returns the default ObjectLibrary
   static std::shared_ptr<ObjectLibrary>& Default();
 
+ protected:
+  friend class ObjectRegistry;
+  // ** FactoryFunctions for this loader, organized by type
+  std::unordered_map<std::string, std::vector<std::unique_ptr<Entry>>> entries_;
+  // Returns the
+  std::string AsPrintableOptions(const std::string& name) const;
+
  private:
   // Adds the input entry to the list for the given type
   void AddEntry(const std::string& type, std::unique_ptr<Entry>& entry);
-
-  // ** FactoryFunctions for this loader, organized by type
-  std::unordered_map<std::string, std::vector<std::unique_ptr<Entry>>> entries_;
 };
 
 // The ObjectRegistry is used to register objects that can be created by a
 // name/pattern at run-time where the specific implementation of the object may
 // not be known in advance.
-class ObjectRegistry {
+class ObjectRegistry : public Configurable {
  public:
   static std::shared_ptr<ObjectRegistry> NewInstance();
 
-  ObjectRegistry();
+  // Makes a copy of this registry.
+  virtual std::shared_ptr<ObjectRegistry> Clone() const = 0;
 
-  void AddLibrary(const std::shared_ptr<ObjectLibrary>& library) {
-    libraries_.emplace_back(library);
-  }
+  // Creates a new local library, registering the factories in registrar
```

Comment:
What's the difference between local library and dynamic library? May need some explanations and examples.

Reply:
I will add a comment later.

A dynamic library is one that was loaded into the application explicitly.  Methods are located by name in that external library.

A "local library" searches the current executable for the method in question, if enabled/allowed.  Methods in a "local library" could be searched by name (if allowed) or their address might be known.  In the latter case, the local library could not be reconstructed on restart.

If you have a better name for the "local library", feel free to suggest one.